### PR TITLE
robot-name ena's solution

### PR DESCRIPTION
**Pull request by enarha**

### CODE REVIEW

File: robot-name/ena/robot_name.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import (
+	"errors"
+	"math"
+	"math/rand"
+	"time"
+)
+
+var usedNames = make(map[string]bool)
+
+func init() {
```
Comment (davidmogar) [inline robot-name/ena/robot_name.go:12]: Nice.

File: robot-name/ena/robot_name.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	rand.Seed(time.Now().UnixNano())
+}
+
+type Robot struct {
+	name string
```
Comment (davidmogar) [inline robot-name/ena/robot_name.go:17]: I would export the name or hide the struct itself.

File: robot-name/ena/robot_name.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	const letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+	const lettersPartLen = 2
+	const digits = "0123456789"
+	const digitsPartLen = 3
+
+	if r.name != "" {
+		return r.name, nil
+	}
+
+	allPossibleNames := int(math.Pow(float64(len(letters)), lettersPartLen) * math.Pow(float64(len(digits)), digitsPartLen))
```
Comment (davidmogar) [inline robot-name/ena/robot_name.go:30]: This should be a constant.

File: robot-name/ena/robot_name.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+	if len(usedNames) == allPossibleNames {
+		return "", errors.New("All names taken")
```
Comment (davidmogar) [inline robot-name/ena/robot_name.go:32]: Errors in go start with a lowercase.

File: robot-name/ena/robot_name.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	}
+	return r.name, nil
+}
+
+func (r *Robot) Reset() {
+	r.name = ""
+	r.Name()
+}
+
+func GenRandomString(pool string, length int) string {
```
Comment (davidmogar) [inline robot-name/ena/robot_name.go:50]: I'd say that this method shouldn't be exported.

File: robot-name/ena/robot_name.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+	if r.name != "" {
+		return r.name, nil
+	}
+
+	allPossibleNames := int(math.Pow(float64(len(letters)), lettersPartLen) * math.Pow(float64(len(digits)), digitsPartLen))
+	if len(usedNames) == allPossibleNames {
+		return "", errors.New("All names taken")
+	}
+
+	for {
```
Comment (davidmogar) [inline robot-name/ena/robot_name.go:35]: This is a costly approach. For your code, the last test had to regenerate 8110729 names as they one generated it already existed. It's still nice to see that is quite fast, so well done.

File: robot-name/ena/robot_name.go (hunk 3 above)
Comment (enarha) [reply robot-name/ena/robot_name.go:30]: I tried to make it const initially, but it fails with the following error `./robot_name.go:30:8: const initializer int(math.Pow(float64(len(letters)), lettersPartLen) * math.Pow(float64(len(digits)), digitsPartLen)) is not a constant`. Did not spend time to understand why.

Comment (davidmogar) [reply robot-name/ena/robot_name.go:30]: You cannot use math.Pow in constants (mostly function calls):

> A constant value is represented by a rune, integer, floating-point, imaginary, or string literal, an identifier denoting a constant, a constant expression, a conversion with a result that is a constant, or the result value of some built-in functions such as unsafe.Sizeof applied to any value, cap or len applied to some expressions, real and imag applied to a complex constant and complex applied to numeric constants. 

You can though hardcode the result of the operation.

File: robot-name/ena/robot_name.go (hunk 2 above)
Comment (enarha) [reply robot-name/ena/robot_name.go:17]: Why? It's like class with private fields managed by getter/setter methods.

Comment (davidmogar) [reply robot-name/ena/robot_name.go:17]: Actually scratch my comment. I don't know what I was thinking about. You have a Name function.

**Pull request merged**